src/Compilers/VisualBasic/Portable/Syntax/SyntaxFactory.vb

```diff
@@ -532,5 +532,64 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
             Return SyntaxEquivalence.AreEquivalent(oldList.Node, newList.Node, ignoreChildNode, topLevel:=False)
         End Function
 
+        ''' <summary>
+        ''' Determites if a submission is complete.
+        ''' Returns false if the syntax is valid but incomplete.
+        ''' Returns true if the syntax is invalid or complete.
+        ''' </summary>
+        ''' <param name="tree">Syntax tree.</param>
+        Public Shared Function IsCompleteSubmission(tree As SyntaxTree) As Boolean
+            Dim options As VisualBasicParseOptions = DirectCast(tree.Options, VisualBasicParseOptions)
+            Dim languageVersion As LanguageVersion = options.LanguageVersion
+
+            If (tree Is Nothing) Then
+                Throw New ArgumentNullException(NameOf(tree))
+            End If
+
+            If (Not tree.HasCompilationUnitRoot) Then
+                Return False
+            End If
+
+            Dim compilation As CompilationUnitSyntax = DirectCast(tree.GetRoot(), CompilationUnitSyntax)
+            If Not compilation.HasErrors Then
+                Return True
+            End If
+
+            For Each err In compilation.GetDiagnostics()
+                Select Case DirectCast(err.Code, ERRID)
+                    Case ERRID.ERR_LbExpectedEndIf,
+                         ERRID.ERR_ExpectedEndRegion
+                        Return False
+                    Case ERRID.ERR_ExpectedEOS
```

Comment:
`ExpectedEOS` diagnostic occurs when two statements were not separated. For example `If condition statement` instead of `If condition Then statement` or `Dim x = 1 Dim y = 2` instead of `Dim x = 1: Dim y = 2`.
What happens for the `If` statement is that the statement is both invalid and incomplete. The case above ensures that in these examples a statement will be reported as invalid rather than incomplete.